modin/pandas/general.py

```diff
@@ -404,27 +404,35 @@ def concat(
                 sort=sort,
             )
         )
-    if join not in ["inner", "outer"]:
+    if join == "outer":
+        # Filter out empties
+        list_of_objs = [
+            obj
+            for obj in list_of_objs
+            if (
+                isinstance(obj, (Series, pandas.Series))
+                or (
+                    isinstance(obj, (DataFrame, Series))
+                    and obj._query_compiler.lazy_execution
+                )
+                or sum(obj.shape) > 0
+            )
+        ]
+    elif join != "inner":
         raise ValueError(
             "Only can inner (intersect) or outer (union) join the other axis"
         )
     # We have the weird Series and axis check because, when concatenating a
     # dataframe to a series on axis=0, pandas ignores the name of the series,
     # and this check aims to mirror that (possibly buggy) functionality
     list_of_objs = [
-        obj
+        obj._query_compiler
         if isinstance(obj, DataFrame)
-        else DataFrame(obj.rename())
+        else DataFrame(obj.rename())._query_compiler
         if isinstance(obj, (pandas.Series, Series)) and axis == 0
-        else DataFrame(obj)
+        else DataFrame(obj)._query_compiler
         for obj in list_of_objs
     ]
-    list_of_objs = [
-        obj._query_compiler
-        for obj in list_of_objs
-        if (not obj._query_compiler.lazy_execution and len(obj.index))
-        or len(obj.columns)
```

Comment:
Wouldn't it be enough to change this condition?
```python
or len(obj.columns) if join == "outer" else True
```
In order not to filter dataframes for the inner join in which there is not an empty index, but empty columns?

Reply:
Probably, it's enough, but, it this case, we just create a copy of the list. I think, it's better to avoid filtering and creating a new list in case of the inner join.

Reply:
Isn't it safe to filter out objects that have an empty index **and** empty columns for an inner join?

Reply:
It's not. For example, we filter out an empty frame and pass down, to the backend, non-empty frames. The backend concatenates the non-empty frames, but, in case of the inner join, the result should be an empty frame with merged index.

Reply:
Got it, thanks for clarification!

Reply:
Pretty simple reproducer:
```python
import modin.pandas as pd
df1 = pd.DataFrame(index=[1,2,3])
df2 = pd.DataFrame(index=[])
pd.concat((df1, df2), axis=1, join="inner")
```